src/main/java/teammates/ui/webapi/GetHasResponsesAction.java

```diff
@@ -55,11 +59,25 @@ void checkSpecificAccessControl() {
         } else {
             //An student can check whether he has submitted responses for a feedback session in his course.
             String courseId = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ID);
-            String feedbackSessionName = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
+            String feedbackSessionName = getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
+            if (feedbackSessionName == null) {
+                List<FeedbackSessionAttributes> feedbackSessions = logic.getFeedbackSessionsForCourse(courseId);
+                if (feedbackSessions.isEmpty()) {
+                    // Course has no sessions and therefore no response; access to responses is safe for all.
+                    return;
+                }
 
-            gateKeeper.verifyAccessible(
-                    logic.getStudentForGoogleId(courseId, userInfo.getId()),
-                    getNonNullFeedbackSession(feedbackSessionName, courseId));
+                // Verify that all sessions are accessible to the user.
+                for (FeedbackSessionAttributes feedbackSession : feedbackSessions) {
```

Comment:
I think we should just ignore unaccessible sessions, the rest can be shown to users

Reply:
Interesting take here: in the access control, we can't do anything but pouring all sessions into access verification; what you are proposing is ditching access verification altogether. Might need to learn about the access control within a course across sessions.

Reply:
Taking a look at `gateKeeper.verifyAccessible`, it appears that a session is inaccessible to a student if the student is not in the course of the session, or the session is not visible.

The former doesn't matter as when we query sessions by course id, if any session is inaccessible to the student by this criterion, all the rest of the sessions shouldn't be accessed anyways; hence, we only need to take care of the latter by adding lines to skip the invisible sessions altogether.

Reply:
I see. Is it possible to extend or modify `gateKeeper.verifyAccessible` to have more fine-grained verification in this case?

Reply:
Can clarify what do you mean by having a fine-grained verification?

Reply:
Right now we are throwing `UnauthorizedAccessException` for both cases of student not being in the course and the session not being visible. I'm proposing a separation of the 2 cases i.e having `NotInCourseException` and `NotVisibleSessionException` so we can return visible sessions and ignore sessions that cause the `NotVisibleException`

Reply:
Was thinking of the same thing and it might be doable; regardless, `gateKeeper.verifyAccesible` is still more coupled with `hasResponseAction` in that either `hasResponseAction` has to know about the access control logic behind `verifyAccessible` or `hasResponseAction` has to know about all the exceptions `verifyAccessible` will throw and treat them accordingly.